### PR TITLE
feat: support resolve_timeout and http_config  alertmanager configuration's global config

### PR DESCRIPTION
**Pull request by jojotong**
## Description

https://github.com/prometheus-operator/prometheus-operator/issues/4606 disscussed global config's implemention, this PR support the following global configs:

1. resolve timeout
2. http config

It will looks like:
```yaml
apiVersion: monitoring.coreos.com/v1
kind: Alertmanager
metadata:
  name: main
spec:
  alertmanagerConfiguration:
    name: main
    global:
      resolveTimeout: 5m
      httpConfig: 
        basicAuth:
          username: uname
          password: passwd 
        oauth2: {}
        ...
```

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

**Comment by jojotong** [comment]
@simonpasquier I force pushed to resolve confilict, and doc in https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/alerting.md#specify-global-alertmanager-config may be finished after #4606 are totally resolved.

File: pkg/alertmanager/amcfg.go [hunk 1]
```diff
@@ -140,44 +141,67 @@ func validateConfigInputs(am *monitoringv1.Alertmanager) error {
 
 func (cg *configGenerator) generateGlobalConfig(
 	ctx context.Context,
+	am *monitoringv1.Alertmanager,
```
[the remark this answers is not in the document]
Comment (jojotong) [reply pkg/alertmanager/amcfg.go:144]: Yes, but we need crKey, maybe I can pass both of crKey and globalConfig to replace the full Alertmanager object.

File: pkg/alertmanager/amcfg.go [hunk 2]
```diff
@@ -140,44 +141,67 @@ func validateConfigInputs(am *monitoringv1.Alertmanager) error {
...
+		if global.HTTPConfig != nil {
+			httpConfig, err := cg.convertHTTPConfigForV1(ctx, *global.HTTPConfig, crKey)
+			if err != nil {
+				return nil, errors.Wrap(err, "convert global http config")
+			}
+			out.Global.HTTPConfig = httpConfig
+		}
+
+		if global.ResolveTimeout != "" {
+			timeout, err := model.ParseDuration(global.ResolveTimeout)
```
[the remark this answers is not in the document]
Comment (jojotong) [reply pkg/alertmanager/amcfg.go:166]: `operator.ValidateDurationField()` only check the duration is parsable or not. I want to pass the parsed duration to `out.ResolveTimeout`.

File: test/e2e/alertmanager_test.go [hunk 3]
```diff
@@ -1364,13 +1364,79 @@ func testUserDefinedAlertmanagerConfigFromCustomResource(t *testing.T) {
 	}
 	alertmanager.Spec.AlertmanagerConfiguration = &monitoringv1.AlertmanagerConfiguration{
 		Name: alertmanagerConfig.Name,
+		Global: &monitoringv1.GlobalConfig{
```
[the remark this answers is not in the document]
Comment (jojotong) [reply test/e2e/alertmanager_test.go:1367]: I'll keep it like this, if we think it's no use, we can remove those fields before this PR be merged.

File: pkg/alertmanager/amcfg.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/alertmanager/amcfg.go]: if the value can be parsed, we can just pass it directly?

[the remark this answers is not in the document]
Comment (jojotong) [reply pkg/alertmanager/amcfg.go]: We can't pass it directly, the raw type of `alertmanagerConfig.Global.ResolveTimeout` is `*model.Duration`.Do you mean change it to `string`?

**Comment by jojotong** [comment]
@simonpasquier pushed again to resolve conflict.

File: pkg/alertmanager/amcfg.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jojotong) [reply pkg/alertmanager/amcfg.go]: ok, changed and rebased

**Comment by jojotong** [comment]
@simonpasquier ping :)

**Comment by ap-vishal** [comment]
> @simonpasquier ping :)

@simonpasquier Could you please help here?

**Comment by jojotong** [comment]
@simonpasquier pushed again to resolve conflicts.

File: pkg/apis/monitoring/v1/types.go [hunk 4]
```diff
@@ -1885,6 +1885,52 @@ type AlertmanagerConfiguration struct {
...
+	Global *AlertmanagerGlobalConfig `json:"global,omitempty"`
+}
+
+// AlertmanagerGlobalConfig configures parameters that are valid in all other configuration contexts.
+// See https://prometheus.io/docs/alerting/latest/configuration/#configuration-file
+type AlertmanagerGlobalConfig struct {
+	// ResolveTimeout is the default value used by alertmanager if the alert does
+	// not include EndsAt, after this time passes it can declare the alert as resolved if it has not been updated.
+	// This has no impact on alerts from Prometheus, as they always include EndsAt.
+	ResolveTimeout string `json:"resolveTimeout,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:1899]:
```suggestion
	ResolveTimeout Duration `json:"resolveTimeout,omitempty"`
```

File: pkg/alertmanager/types.go [hunk 5]
```diff
@@ -39,7 +38,7 @@ type alertmanagerConfig struct {
 type globalConfig struct {
 	// ResolveTimeout is the time after which an alert is declared resolved
 	// if it has not been updated.
-	ResolveTimeout *model.Duration `yaml:"resolve_timeout,omitempty" json:"resolve_timeout,omitempty"`
```
Comment (simonpasquier) [inline pkg/alertmanager/types.go:42]: not sure why you want to change it

Comment (jojotong) [reply pkg/alertmanager/types.go]: I forget too, but fixed

**Review by simonpasquier — APPROVED** [review]
thanks!

**Pull request merged**